tools/java/org/unicode/cldr/util/DtdData.java

```diff
@@ -654,20 +653,47 @@ public void handleEndDtd() {
     /**
      * Normal version of DtdData
      * Note that it always gets the trunk version
+     * @deprecated depends on static configuration (CLDR_DIR)
      */
     public static DtdData getInstance(DtdType type) {
-        return CACHE.get(type);
+        return getInstance(type, CLDRConfig.getInstance().getCldrBaseDirectory());
     }
-
+    
     /**
      * Special form using version, used only by tests, etc.
      */
     public static DtdData getInstance(DtdType type, String version) {
-        DtdData simpleHandler = new DtdData(type, version);
-        XMLFileReader xfr = new XMLFileReader().setHandler(simpleHandler);
         File directory = version == null ? CLDRConfig.getInstance().getCldrBaseDirectory()
             : new File(CLDRPaths.ARCHIVE_DIRECTORY + "/cldr-" + version);
 
+        return getInstance(type, version, directory);
+    }
+    
+    private static final Map<Pair<DtdType, File>, DtdData> OTHERCACHE = new  HashMap<>();
+
+    /**
+     * Get a DtdType, given a CLDR root directory.
+     * @param type
+     * @param directory
+     * @return
+     */
+    public static synchronized DtdData getInstance(DtdType type, File directory) {
```

Comment:
I'd suggest ConcurrentMap and "computeIfAbsent()" since that avoid any risk of someone mistakenly reading from the map outside a locked section and won't hold a look during the expensive loading of the DtdData instance.

If you are synchronizing on something, I'd suggest not using the static class instance (which is what static synchonrized uses). I know the synchronized keyword advertises the synchronization to callers, but there's an argument that this is implementation detail (e.g. it becomes a visible API change to switch to using a concurrent map now). Perhaps just synchronize on the private cache instance itself?

Reply:
Good suggestions, I'll do this.

Reply:
if i use ConcurrentMap/putIfAbsent, I think synchronized is not needed here.

Reply:
^PTAL